// @mailchain/crypto/node/cipher/index.js
'use strict';

var ed25519 = require('./ecdh/ed25519.js');
var secp256k1 = require('./ecdh/secp256k1.js');
var sr25519 = require('./ecdh/sr25519.js');
var cipher = require('./cipher.js');
var publicKeyEncrypter = require('./nacl/publicKeyEncrypter.js');
var publicKeyDecrypter = require('./nacl/publicKeyDecrypter.js');
var privateKeyEncrypter = require('./nacl/privateKeyEncrypter.js');
var privateKeyDecrypter = require('./nacl/privateKeyDecrypter.js');
var ecdh = require('./ecdh/ecdh.js');



exports.ED25519KeyExchange = ed25519.ED25519KeyExchange;
exports.SECP256K1KeyExchange = secp256k1.SECP256K1KeyExchange;
exports.SR25519KeyExchange = sr25519.SR25519KeyExchange;
exports.KindNaClSecretKey = cipher.KindNaClSecretKey;
exports.NACLECDH = cipher.NACLECDH;
exports.NACLSK = cipher.NACLSK;
exports.PublicKeyEncrypter = publicKeyEncrypter.PublicKeyEncrypter;
exports.PublicKeyDecrypter = publicKeyDecrypter.PublicKeyDecrypter;
exports.PrivateKeyEncrypter = privateKeyEncrypter.PrivateKeyEncrypter;
exports.PrivateKeyDecrypter = privateKeyDecrypter.PrivateKeyDecrypter;
exports.fromPrivateKey = ecdh.fromPrivateKey;
exports.fromPublicKey = ecdh.fromPublicKey;
